### PR TITLE
Add support for Boolector as a backend engine for Kind2.

### PR DESCRIPTION
**Pull request by abdoo8080**
This PR adds support for Boolector (an SMT solver) for Lustre problems with only machine integers. Since Boolector does not support IA/RA, an error is printed when those sub-logics are detected and Boolector is selected as the backend engine. Techniques using IA/RA are either disabled or modified to work with BV too. Finally, QF_UF logic is expanded to QF_UFBV as the former is not supported by Boolector.

### CODE REVIEW

File: src/flags.ml [hunk 1]
```diff
@@ -117,25 +117,28 @@ module Smt = struct
...
     | _ -> Arg.Bad "Bad value for --smt_solver" |> raise
   let string_of_solver = function
+    | `Boolector_SMTLIB -> "Boolector"
     | `Z3_SMTLIB -> "Z3"
     | `CVC4_SMTLIB -> "CVC4"
     | `Yices_SMTLIB -> "Yices2"
     | `Yices_native -> "Yices"
     | `detect -> "detect"
-  let solver_values = "Z3, CVC4, Yices, Yices2"
+  let solver_values = "Boolector, Z3, CVC4, Yices, Yices2"
```
Comment (daniel-larraz) [inline src/flags.ml:141]: Use the order of detection (Z3, Yices2, CVC4, Yices, Boolector)

File: src/lustre/lustreExpr.ml [hunk 2]
```diff
@@ -1778,6 +1778,10 @@ let mk_to_int expr = mk_unary eval_to_int type_of_to_int expr
 
 (* ********************************************************************** *)
 
+let is_negative term = match Term.destruct term with
+  | Term.T.App (s, _) when s == Symbol.s_minus -> true
+  | _ -> false
+
```
Comment (daniel-larraz) [inline src/lustre/lustreExpr.ml:1784]: Move function to the Term module

File: src/smt/BoolectorDriver.ml [hunk 3]
```diff
@@ -0,0 +1,71 @@
+(* This file is part of the Kind 2 model checker.
+
+   Copyright (c) 2015 by the Board of Trustees of the University of Iowa
```
Comment (daniel-larraz) [inline src/smt/BoolectorDriver.ml:3]: Change year to 2020

File: src/terms/termLib.ml [hunk 4]
```diff
@@ -256,19 +256,17 @@ type logic = [ `None | `Inferred of features | `SMTLogic of string ]
...
 let string_of_logic = function
   | `None -> "ALL"
-  | `Inferred l -> if (L.mem BV l) then 
-                    "ALL"
-                   else
-                    string_of_features l
+  | `Inferred l ->
+      if L.mem BV l && (L.mem IA l || L.mem RA l) then "ALL"
+      else string_of_features l
   | `SMTLogic s -> if s = "" then "ALL" else s
```
Comment (daniel-larraz) [inline src/terms/termLib.ml:270]: Define this function in terms of `pp_print_logic` to synchronize logic:
`let string_of_logic l = asprintf "%a" pp_print_logic l`

File: src/flags.ml (hunk 1 above)
Comment (abdoo8080) [reply src/flags.ml]: Done!

File: src/smt/BoolectorDriver.ml (hunk 3 above)
Comment (abdoo8080) [reply src/smt/BoolectorDriver.ml]: Done!

File: src/terms/termLib.ml (hunk 4 above)
Comment (abdoo8080) [reply src/terms/termLib.ml]: Done!

File: src/lustre/lustreExpr.ml (hunk 2 above)
Comment (abdoo8080) [reply src/lustre/lustreExpr.ml]: Done!

**Pull request merged**